signer/src/block_notifier.rs

```diff
@@ -0,0 +1,150 @@
+//! This module provides functionality for notifying about new blocks using the Electrum client.
+
+use crate::config::SETTINGS;
+use electrum_client::bitcoin::BlockHash;
+use electrum_client::{
+    Client, ConfigBuilder, ElectrumApi, Error as ElectrumError, HeaderNotification,
+};
+use futures::stream::Stream;
+use std::error::Error;
+use std::pin::Pin;
+use std::sync::Arc;
+use tokio::sync::mpsc::{channel, Sender};
+use tokio::task;
+use tokio::time::{interval, sleep, Duration};
+use tokio_stream::wrappers::ReceiverStream;
+use tracing::{error, warn};
+
+/// The `BlockNotifier` trait defines a method for subscribing to a stream of block headers.
+pub trait BlockNotifier {
+    /// Errors occurring during subscription or running the notifier.
+    type Error: Error;
+
+    /// Returns a stream of block headers.
+    fn subscribe(&self) -> Pin<Box<dyn Stream<Item = Result<BlockHash, Self::Error>> + Send>>;
+}
+
+/// A struct implementing the `BlockNotifier` trait using Electrum client.
+pub struct ElectrumBlockNotifier {
+    client: Arc<Client>,
+    retry_interval: Duration,
+    max_retry_attempts: u32,
+    ping_interval: Duration,
+    subscribe_interval: Duration,
+}
+
+impl ElectrumBlockNotifier {
+    /// Creates a new instance of `ElectrumBlockNotifier` from config settings.
+    ///
+    /// # Returns
+    ///
+    /// A new instance of `ElectrumBlockNotifier`.
+    pub fn from_config() -> Result<Self, ElectrumError> {
```

Comment:
Can we take the config as an argument here? In the binary we can just inject the system config, but for test cases I'd like the test configuration to be explicit and not dependent on the system environment if possible.